design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,989 @@
+  classSelector:
+    matchLabels:
+      resourceclass.crossplane.io/aggregate-to: cool-gke-cluster
+specTemplate:
+  writeConnectionSecretsToNamespace: crossplane-system
+```
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResource
+metadata:
+  name: default-coolcluster-g3bf7
+spec:
+  aggregationRule:
+    resourceSelector:
+      matchLabels:
+        resourceclaim.crossplane.io/namespace: default
+        resourceclaim.crossplane.io/name: coolcluster
+        resourceclaim.crossplane.io/uid: eabce854-0cd7-11ea-8d71-362b9e155667
+  claimRef:
+    apiVersion: compute.crossplane.io/v1alpha1
+    kind: KubernetesCluster
+    namespace: default
+    name: coolcluster
+    uid: eabce854-0cd7-11ea-8d71-362b9e155667
+  classRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResourceClass
+    name: cool-gke-cluster
+    uid: ea414c74-0cd9-11ea-8d71-362b9e155667
+  writeConnectionSecretsToRef:
+    namespace: crossplane-system
+    name: eabce854-0cd7-11ea-8d71-362b9e155667
+```
+
+The purpose of these new resource kinds, as their names imply, is to _aggregate_
+resource classes and managed resources, enabling transitive one-to-many
+resource-claim-to-resource-class and resource-claim-to-managed-resource
+relationships. Both new kinds use an _aggregation rule_, like an [RBAC
+`ClusterRole`]. The `AggregateResourceClass` aggregates resource classes (of any
+non-aggregate kind) by label, while the `AggregateResource` aggregates managed
+resources (again, of any non-aggregate kind) by label.
+
+In the above example `default-coolcluster-g3bf7` is an `AggregateResource` that
+was dynamically provisioned to satisfy a `KubernetesCluster` resource claim. An
+`AggregateResource` was called for because the `KubernetesCluster` specified an
+`AggregateResourceClass` as its `classRef`:
+
+```yaml
+apiVersion: compute.crossplane.io/v1beta1
+kind: KubernetesCluster
+metadata:
+  namespace: default
+  name: coolcluster
+spec:
+  classRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResourceClass
+    name: cool-gke-cluster
+```
+
+The creation of the above `KubernetesCluster` claim triggers:
+
+1. The dynamic provisioning of the `AggregateResource`, which aggregates all
+   managed resources labelled with the claim's namespace, name, and UID.
+1. The dynamic provisioning of a managed resource for each resource class that
+   aggregates to the `AggregateResourceClass` by matching its `aggregationRule`
+   labels. These dynamically provisioned managed resources are labelled with the
+   claim's namespace, name, and UID and thus aggregate to the dynamically
+   provisioned `AggregateResource`.
+
+The `AggregateResource` binds to the `KubernetesCluster` claim, while the
+dynamically provisioned managed resources bind in turn to the
+`AggregateResource`, thus creating a `resource claim -> aggregate resource ->
+managed resource(s)` binding relationship. The below `GKECluster` is
```

Comment:
Bindings are bidirectional - they both bind to each other. So this would probably be more accuratel as `resource claim <-> aggregate resource <->
managed resource(s)`